### DISCUSSION TITLE
Allow later var variable type detection

### DISCUSSION THREAD

**Discussion by wanton7** [dotnet/csharplang#448] [General]
Later type detection for variable defined with var should be allowed. Following should work
```
var value;
lock (_obj)
{
   value = GetValue();
}
```
and
```
var value = null;
if (condition)
{
  value = GetValue();
}
```

**Comment by HaloFour** [comment]
I think it would make the code a lot less readable to have to scan through the code to find an eventual initialization to determine the type of the variable.

Given that `value` is mutable when would you suggest the compiler determine the type given multiple assignments?  What about if there are assignments in different branches of different types?  For example, what if `GetValue()` returns `int` in your second case?

**Comment by wanton7** [comment]
@HaloFour for second case compiler error if GetValue returns int or something that isn't nullable.

For multiple assingments it would work like ? : does now. Meaning types would have to be compatible. 

Visual Studio already shows var variables type if put mouse pointer top of it. So no need to scan through code.

**Comment by DavidArno** [comment]
You first example would be addressed with [lock and using expressions](https://github.com/dotnet/csharplang/issues/451):
```cs
var value = lock(_obj) GetValue();
```
Your second example is already handled with a ternary expression:
```cs
var value = condition ? GetValue() : null;
```

**Comment by jcouv** [comment]
Relates to https://github.com/dotnet/csharplang/issues/520 (allow `var x;` with inference from later usage)
